### DISCUSSION TITLE
Transfer material parameters to a Sub-App (MultiApps)

### DISCUSSION THREAD

**Discussion by LabrosV** [idaholab/moose#18104] [Q&A General] [answered]
Hi all,

In my main app, I declare a Material Property, which is different at each quadrature point (and can be considered as a history variable). 

When I needed to use a staggered scheme between 2 coupled fields, I used the MultiApp system to do so. The aforementioned history variable is requested in the sub app but it cannot be transferred based on other posts I read here. I managed to convert this Material Property into AuxVariable (that can be transferred) using MaterialRealAux but this approach makes this history variable constant throughout the element in the sub-app.

Questions:
1) How can I convert this material property into AuxVariable with higher order (not constant)? Please consider that I want to be able to run my code in a cluster (so higher order approximations at element level (i.e. the MONOMIAL option with high order) cannot be used because they cannot run in parallel). 

2) I created an AuxKernel which returns an AuxVariable based on the values of the Material Property. However, I think this solution is wrong because this kernel must run on a nodal variable and not material property. Can I use an AuxKernel to convert the data? Any suggestion for a Postprocessor method that I have overlooked?

3) Should I change my implementation in the main app and declare an AuxVariable instead of a Material Property? Then, it will be easier to transfer it.

4) Any alternative ways to implement a staggered scheme without using the MultiApps system? Maybe a field-split solver in the executioner?

Answers to any of these questions are much appreciated.

Thanks,
Labros

**Comment by GiudGiud** [comment, marked as answer]
Hello

Transferring to AuxVariables is the right thing to do here. However, using a higher order AuxVariable and essentially projecting the material property may not have been implemented. @fdkong worked on projections for transfers, which is a similar issue, he should be able to tell us more. 

AuxKernels are the right way to copy MaterialProperties onto an AuxKernel. Postprocessors will only process like one value for the whole domain. 

If that is an option, I think it would work. Then you will avoid the projection issue. 

Field split executioner is an option but it s not really staggered. Please let us know how that goes if you try it. Multiapp is the easier option for staggering. 

best,
Guillaume

**Comment by fdkong** [comment]
>How can I convert this material property into AuxVariable with higher order (not constant)? Please consider that I want to be able to run my code in a cluster (so higher order approximations at element level (i.e. the MONOMIAL option with high order) cannot be used because they cannot run in parallel).

What error messages you saw when using higher order variables?  `MaterialRealAux` should work with  higher order variables and it will do projections.  

  >  I created an AuxKernel which returns an AuxVariable based on the values of the Material Property. However, I think this solution is wrong because this kernel must run on a nodal variable and not material property. Can I use an AuxKernel to convert the data? Any suggestion for a Postprocessor method that I have overlooked?
  
 Yes, but  `MaterialRealAux` should do right things for you. AuxKernels can run on either nodal variables or elemental variables. 

>    Should I change my implementation in the main app and declare an AuxVariable instead of a Material Property? Then, it will be easier to transfer it.

This option will work, but it is not necessary if your code already with materials.

 >   Any alternative ways to implement a staggered scheme without using the MultiApps system? Maybe a field-split solver in the executioner?
 
 A field-split solver  is a preconditioner, and it does not fit well for your purpose here.
 
 However, even you have a high order variable using, and the current implementation of transfers does not work well with high order variables yet. We have a PR here https://github.com/idaholab/moose/pull/17417 for enabling high-order variable transfers
